sa/sa.go

```diff
@@ -2169,3 +2470,245 @@ func (ssa *SQLStorageAuthority) RevokeCertificate(ctx context.Context, req *sapb
+	oldAuthzMap, err := ssa.GetValidOrderAuthorizations(ctx, req)
+	if err != nil {
+		return nil, err
+	}
+	if len(oldAuthzMap) > 0 {
+		oldAuthzsPB, err := authzMapToPB(oldAuthzMap)
+		if err != nil {
+			return nil, err
+		}
+		for _, authzPB := range oldAuthzsPB.Authz {
+			authzsPB.Authz = append(authzsPB.Authz, authzPB)
+		}
+	}
+
+	return authzsPB, nil
+}
+
+// CountInvalidAuthorizations2 counts invalid authorizations for a user expiring
+// in a given time range. This method is intended to deprecate CountInvalidAuthorizations.
+func (ssa *SQLStorageAuthority) CountInvalidAuthorizations2(ctx context.Context, req *sapb.CountInvalidAuthorizationsRequest) (*sapb.Count, error) {
+	var count int64
+	err := ssa.dbMap.WithContext(ctx).SelectOne(
+		&count,
+		`SELECT COUNT(1) FROM authz2 WHERE
+		registrationID = :regID AND
+		identifierValue = :ident AND
+		expires > :expiresEarliest AND
+		expires <= :expiresLatest AND
+		status = :status`,
+		map[string]interface{}{
+			"regID":           *req.RegistrationID,
+			"ident":           *req.Hostname,
+			"expiresEarliest": time.Unix(0, *req.Range.Earliest),
+			"expiresLatest":   time.Unix(0, *req.Range.Latest),
+			"status":          statusUnit(core.StatusInvalid),
+		},
+	)
+	if err != nil {
+		return nil, err
+	}
+	// Also count old authorizations and add those to the new style
+	// style count
+	oldCount, err := ssa.CountInvalidAuthorizations(ctx, req)
+	if err != nil {
+		return nil, err
+	}
+	count += *oldCount.Count
+	return &sapb.Count{Count: &count}, nil
+}
+
+// GetValidAuthorizations2 returns the latest authorization for all
+// domain names that the account has authorizations for. This method is
+// intended to deprecate GetValidAuthorizations.
+func (ssa *SQLStorageAuthority) GetValidAuthorizations2(ctx context.Context, req *sapb.GetValidAuthorizationsRequest) (*sapb.Authorizations, error) {
+	qmarks := make([]string, len(req.Domains))
+	for i := range req.Domains {
+		qmarks[i] = "?"
+	}
+
+	var authzModels []authz2Model
+	params := []interface{}{
+		*req.RegistrationID,
+		time.Unix(0, *req.Now),
+		statusUnit(core.StatusValid),
+	}
+	for _, n := range req.Domains {
+		params = append(params, n)
+	}
+	_, err := ssa.dbMap.Select(
+		&authzModels,
+		fmt.Sprintf(
+			`SELECT %s from authz2 WHERE
+			registrationID = ? AND
+			expires > ? AND
+			status = ? AND
+			identifierValue IN (%s)`,
+			authz2Fields,
+			strings.Join(qmarks, ","),
+		),
+		params...,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	authzMap := make(map[string]authz2Model, len(authzModels))
+	for _, am := range authzModels {
+		authzMap[am.IdentifierValue] = am
+	}
+	authzsPB, err := authz2ModelMapToPB(authzMap)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(authzsPB.Authz) != len(req.Domains) {
+		// We may still have valid old style authorizations
+		// we want for names in the list, so we have to look
+		// for them.
+		leftOverMap := make(map[string]bool, len(req.Domains))
```

Comment:
Similar to my comment above, we don't have to build a map and then delete from it. Instead we can build a slice out of only those domains that are missing from the `authzMap`.

By the way, `GetAuthorizations2` has  a lot of code in common with `GetValidAuthorizations2`. Can we combine them to reduce duplication?

Reply:
While they are similarish I think there are enough subtle differences in how they operate that attempting to combine them, or even pull out any similar functionality into helpers, would actually make the resulting method for more complicated and hard to reason about than what we currently have.